clippy_lints/src/xor_used_as_pow.rs

```diff
@@ -0,0 +1,79 @@
+use crate::utils::{span_lint_and_help, span_lint_and_sugg};
+use if_chain::if_chain;
+use rustc_ast::{BinOpKind, Expr, ExprKind, LitKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for use of `^` operator when exponentiation was intended.
+    ///
+    /// **Why is this bad?** This is most probably a typo.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// // Bad
+    /// 2 ^ 16;
+    ///
+    /// // Good
+    /// 1 << 16;
+    /// 2i32.pow(16);
+    /// ```
+    pub XOR_USED_AS_POW,
+    correctness,
+    "use of `^` operator when exponentiation was intended"
+}
+
+declare_lint_pass!(XorUsedAsPow => [XOR_USED_AS_POW]);
+
+impl EarlyLintPass for XorUsedAsPow {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
```

Comment:
Okay, that makes sense. When checking `Expr`s, is there some easy way to see if we're in an enum discriminant?

Reply:
Since you have to turn your pass into a `LateLintPass`, you can crawl the syntax tree in the opposite direction with [cx.tcx.hir().get_parent_item()](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.get_parent_item) and see if you find an [Enum](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.ItemKind.html#variant.Enum).